File: lib/foundation/dependency_identity.ml
type classification = Local | Immutable | Mutable | Unknown

let is_hex = function
  | '0' .. '9' | 'a' .. 'f' | 'A' .. 'F' -> true
  | _ -> false

let exact_hex length value =
  String.length value = length && String.for_all is_hex value

let valid_content_digest value =
  let prefix = "sha256:" in
  String.length value = String.length prefix + 64
  && Util.starts_with ~prefix value
  && exact_hex 64 (String.sub value (String.length prefix) 64)

let immutable_revision revision =
  exact_hex 40 revision || exact_hex 64 revision
  || valid_content_digest revision

let classify_reference reference =
  if
    Util.starts_with ~prefix:"./" reference
    || Util.starts_with ~prefix:"../" reference
  then Local
  else
    match String.rindex_opt reference '@' with
    | Some index ->
        let revision =
          String.sub reference (index + 1) (String.length reference - index - 1)
        in
        if immutable_revision revision then Immutable else Mutable
    | None -> if Util.contains ~needle:"://" reference then Mutable else Unknown

File: crates/foundation/src/dependency_identity.rs
/// Provider-independent mutability classification.
#[derive(Clone, Copy, Debug, Eq, Ord, PartialEq, PartialOrd)]
pub enum DependencyClass {
    Local,
    Immutable,
    Mutable,
    Unknown,
}

fn exact_hex(value: &str, length: usize) -> bool {
    value.len() == length && value.bytes().all(|byte| byte.is_ascii_hexdigit())
}

fn exact_lower_hex(value: &str, length: usize) -> bool {
    value.len() == length
        && value
            .bytes()
            .all(|byte| byte.is_ascii_digit() || (b'a'..=b'f').contains(&byte))
}

#[must_use]
pub fn valid_content_digest(value: &str) -> bool {
    value
        .strip_prefix("sha256:")
        .is_some_and(|hex| exact_lower_hex(hex, 64))
}

fn immutable_revision(value: &str) -> bool {
    exact_hex(value, 40) || exact_hex(value, 64) || valid_content_digest(value)
}

#[must_use]
pub fn classify_reference(reference: &str) -> DependencyClass {
    if reference.starts_with("./") || reference.starts_with("../") {
        return DependencyClass::Local;
    }
    if let Some((_, revision)) = reference.rsplit_once('@') {
        if immutable_revision(revision) {
            DependencyClass::Immutable
        } else {
            DependencyClass::Mutable
        }
    } else if reference.contains("://") {
        DependencyClass::Mutable
    } else {
        DependencyClass::Unknown
    }
}

#[cfg(test)]
mod tests {
    use super::{DependencyClass, classify_reference, valid_content_digest};

    #[test]
    fn provider_independent_classification() {
        assert_eq!(classify_reference("./action"), DependencyClass::Local);
        assert_eq!(
            classify_reference("org/action@main"),
            DependencyClass::Mutable
        );
        assert_eq!(
            classify_reference("org/action@0123456789abcdef0123456789abcdef01234567"),
            DependencyClass::Immutable
        );
        assert_eq!(
            classify_reference("docker://alpine:latest"),
            DependencyClass::Mutable
        );
        assert_eq!(classify_reference("opaque"), DependencyClass::Unknown);
        assert!(valid_content_digest(&format!("sha256:{}", "a".repeat(64))));
        assert!(!valid_content_digest(&format!("sha256:{}", "g".repeat(64))));
    }
}
